presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java

```diff
@@ -824,17 +889,45 @@ public static Slice roundNLong(
             }
         }
 
+        @LiteralParameters({"p", "s", "rp"})
+        @SqlType("decimal(rp, s)")
+        @Constraint(variable = "rp", expression = "min(38, p + 1)")
+        public static Slice roundNLongBigintDecimals(
+                ConnectorSession session,
+                @LiteralParameter("s") long numScale,
+                @LiteralParameter("rp") long resultPrecision,
+                @SqlType("decimal(p, s)") Slice num,
+                @SqlType(StandardTypes.BIGINT) long decimals)
+        {
+            checkCondition(session.isLegacyRoundNBigint(), INVALID_FUNCTION_ARGUMENT, "ROUND(x, d) does not accept d being a BIGINT. Please cast the second argument to INTEGER");
+            return roundNLong(numScale, resultPrecision, num, toIntExact(decimals));
+        }
+
         @LiteralParameters({"p", "s", "rp"})
         @SqlType("decimal(rp, s)")
         @Constraint(variable = "rp", expression = "min(38, p + 1)")
         public static Slice roundNShortLong(
                 @LiteralParameter("s") long numScale,
                 @LiteralParameter("rp") long resultPrecision,
                 @SqlType("decimal(p, s)") long num,
-                @SqlType(StandardTypes.BIGINT) long decimals)
+                @SqlType(StandardTypes.INTEGER) long decimals)
         {
             return roundNLong(numScale, resultPrecision, unscaledDecimal(num), decimals);
         }
+
+        @LiteralParameters({"p", "s", "rp"})
+        @SqlType("decimal(rp, s)")
+        @Constraint(variable = "rp", expression = "min(38, p + 1)")
+        public static Slice roundNShortLongBigintDecimals(
+                ConnectorSession session,
+                @LiteralParameter("s") long numScale,
+                @LiteralParameter("rp") long resultPrecision,
+                @SqlType("decimal(p, s)") long num,
+                @SqlType(StandardTypes.BIGINT) long decimals)
+        {
+            checkCondition(session.isLegacyRoundNBigint(), INVALID_FUNCTION_ARGUMENT, "ROUND(x, d) does not accept d being a BIGINT. Please cast the second argument to INTEGER");
```

Comment:
use `ROUNDN` in exeception message

Reply:
You mean without `(x, d)`? How would you word this? `ROUNDN does not accept N being a BIGINT. Please cast the second argument to INTEGER` ?

Reply:
Never mind as above.